File: sdk/python/samples/basic_streaming.py
#!/usr/bin/env python3
from __future__ import annotations

from codex_sdk import Codex
from codex_sdk.types import ThreadEvent, ThreadItem

from helpers import codex_path_override


def handle_item_completed(item: ThreadItem) -> None:
    item_type = item.get("type")
    if item_type == "agent_message":
        print(f"Assistant: {item.get('text')}")
    elif item_type == "reasoning":
        print(f"Reasoning: {item.get('text')}")
    elif item_type == "command_execution":
        exit_code = item.get("exit_code")
        exit_text = f" Exit code {exit_code}." if exit_code is not None else ""
        print(f"Command {item.get('command')} {item.get('status')}.{exit_text}")
    elif item_type == "file_change":
        for change in item.get("changes", []):
            print(f"File {change.get('kind')} {change.get('path')}")


def handle_item_updated(item: ThreadItem) -> None:
    if item.get("type") == "todo_list":
        print("Todo:")
        for todo in item.get("items", []):
            mark = "x" if todo.get("completed") else " "
            print(f"\t {mark} {todo.get('text')}")


def handle_event(event: ThreadEvent) -> None:
    event_type = event.get("type")
    if event_type == "item.completed":
        handle_item_completed(event.get("item", {}))
    elif event_type in {"item.updated", "item.started"}:
        handle_item_updated(event.get("item", {}))
    elif event_type == "turn.completed":
        usage = event.get("usage", {})
        print(
            "Used {input} input tokens, {cached} cached input tokens, {output} output tokens.".format(
                input=usage.get("input_tokens"),
                cached=usage.get("cached_input_tokens"),
                output=usage.get("output_tokens"),
            )
        )
    elif event_type == "turn.failed":
        error = event.get("error", {})
        print(f"Turn failed: {error.get('message')}")


def main() -> None:
    codex = Codex(codex_path_override=codex_path_override())
    thread = codex.start_thread()

    try:
        while True:
            user_input = input("> ").strip()
            if not user_input:
                continue
            streamed = thread.run_streamed(user_input)
            for event in streamed.events:
                handle_event(event)
    except KeyboardInterrupt:
        return


if __name__ == "__main__":
    main()

File: sdk/typescript/samples/basic_streaming.ts
#!/usr/bin/env -S NODE_NO_WARNINGS=1 pnpm ts-node-esm --files

import { createInterface } from "node:readline/promises";
import { stdin as input, stdout as output } from "node:process";

import { Codex } from "@openai/codex-sdk";
import type { ThreadEvent, ThreadItem } from "@openai/codex-sdk";
import { codexPathOverride } from "./helpers.ts";

const codex = new Codex({ codexPathOverride: codexPathOverride() });
const thread = codex.startThread();
const rl = createInterface({ input, output });

let running = true;

/* Graceful shutdown */
process.on("SIGINT", () => {
  console.log("\nExiting Codex CLI...");
  running = false;
  rl.close();
  process.exit(0);
});

const handleItemCompleted = (item: ThreadItem): void => {
  switch (item.type) {
    case "agent_message":
      console.log(`Assistant: ${item.text}`);
      break;

    case "reasoning":
      console.log(`Reasoning: ${item.text}`);
      break;

    case "command_execution": {
      const exitText = item.exit_code !== undefined ? ` Exit code ${item.exit_code}.` : "";
      console.log(`Command ${item.command} ${item.status}.${exitText}`);
      break;
    }

    case "file_change": {
      for (const change of item.changes) {
        console.log(`File ${change.kind} ${change.path}`);
      }
      break;
    }

    default:
      // Future-proof: log unknown item types for debugging
      console.debug(`Unhandled completed item type: ${(item as any).type}`);
  }
};

const handleItemUpdated = (item: ThreadItem): void => {
  switch (item.type) {
    case "todo_list": {
      console.log(`Todo:`);
      for (const todo of item.items) {
        console.log(`\t ${todo.completed ? "x" : " "} ${todo.text}`);
      }
      break;
    }

    default:
      console.debug(`Unhandled updated item type: ${(item as any).type}`);
  }
};

const handleEvent = (event: ThreadEvent): void => {
  switch (event.type) {
    case "item.completed":
      handleItemCompleted(event.item);
      break;

    case "item.updated":
    case "item.started":
      if (event.item) handleItemUpdated(event.item);
      break;

    case "turn.completed":
      console.log(
        `Used ${event.usage.input_tokens} input tokens, ${event.usage.cached_input_tokens} cached input tokens, ${event.usage.output_tokens} output tokens.`,
      );
      break;

    case "turn.failed":
      console.error(`Turn failed: ${event.error.message}`);
      break;

    default:
      console.debug(`Unhandled event type: ${(event as any).type}`);
  }
};

const main = async (): Promise<void> => {
  try {
    while (running) {
      const inputText = await rl.question("You > ");
      const trimmed = inputText.trim();

      if (trimmed.length === 0) continue;

      try {
        const { events } = await thread.runStreamed(trimmed);

        for await (const event of events) {
          handleEvent(event);
        }
      } catch (err) {
        const message = err instanceof Error ? err.message : String(err);
        console.error(`Turn error: ${message}`);
      }
    }
  } finally {
    rl.close();
  }
};

main().catch((err) => {
  const message = err instanceof Error ? err.message : String(err);
  console.error(`Unexpected error: ${message}`);
  process.exit(1);
});
